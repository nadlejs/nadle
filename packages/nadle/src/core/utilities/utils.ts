import Path from "node:path";

import c from "tinyrainbow";

import { DOT } from "./constants.js";

export function noop() {}

export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

const MILLISECONDS_PER_SECOND = 1000;
const MILLISECONDS_PER_MINUTE = 60 * MILLISECONDS_PER_SECOND;
const MILLISECONDS_PER_HOUR = 60 * MILLISECONDS_PER_MINUTE;

export function formatTime(ms: number): string {
	const hours = Math.floor(ms / MILLISECONDS_PER_HOUR);
	const minutes = Math.floor((ms % MILLISECONDS_PER_HOUR) / MILLISECONDS_PER_MINUTE);
	const seconds = Math.floor((ms % MILLISECONDS_PER_MINUTE) / MILLISECONDS_PER_SECOND);
	const milliseconds = Math.floor(ms % MILLISECONDS_PER_SECOND);

	if (hours > 0) {
		if (minutes > 0) {
			return `${hours}h${minutes}m`;
		}

		return `${hours}h`;
	}

	if (minutes > 0) {
		if (seconds > 0) {
			return `${minutes}m${seconds}s`;
		}

		return `${minutes}m`;
	}

	if (seconds > 0) {
		return `${seconds}s`;
	}

	return `${milliseconds}ms`;
}

export function normalizeGlobPath(path: string) {
	if (path.startsWith(DOT)) {
		return path;
	}

	return `.${Path.sep}${path}`;
}

export function clamp(value: number, min: number, max: number): number {
	return Math.min(Math.max(value, min), max);
}

export function bindObject<T, K extends keyof T>(obj: T, keys: K[]): { [P in K]: T[P] extends (...args: any[]) => any ? T[P] : never } {
	return Object.fromEntries(
		keys.map((key) => {
			const value = obj[key];

			if (typeof value !== "function") {
				throw new Error(`Property ${String(key)} is not a function`);
			}

			return [key, value.bind(obj)];
		})
	) as any;
}

export function highlight(text: string): string {
	return c.bold(c.yellow(text));
}
