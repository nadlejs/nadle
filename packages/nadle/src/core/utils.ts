export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

export function formatTime(time: number): string {
	if (time > 1000) {
		return `${(time / 1000).toFixed(2)}s`;
	}

	return `${Math.round(time)}ms`;
}

export function formatTimeString(date: Date): string {
	return date.toTimeString().split(" ")[0];
}
