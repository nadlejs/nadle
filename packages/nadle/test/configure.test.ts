import { it, describe } from "vitest";

import { createExec, expectPass } from "./setup/utils.js";

describe("Configure", () => {
	const exec = createExec({ config: "configure" });

	it("can use configured options from config file", async () => {
		await expectPass(exec`--show-config`);
	});

	it("can override configured options from config file if the cli one is provided", async () => {
		await expectPass(exec`--log-level info --show-config`);
	});
});
