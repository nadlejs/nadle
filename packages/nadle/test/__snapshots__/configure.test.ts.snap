// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`configure > can override configured options from config file if the cli one is provided > stdout 1`] = `
[info] Initialized logger with consola options: { types:
   { silent: { level: -1 },
     fatal: { level: 0 },
     error: { level: 0 },
     warn: { level: 1 },
     log: { level: 2 },
     info: { level: 3 },
     success: { level: 3 },
     fail: { level: 3 },
     ready: { level: 3 },
     start: { level: 3 },
     box: { level: 3 },
     debug: { level: 4 },
     trace: { level: 5 },
     verbose: { level: Infinity } },
  throttle: 1000,
  throttleMin: 5,
  formatOptions: { date: false, colors: false, compact: true },
  level: 3,
  defaults: { level: 1 },
  prompt: [Function: prompt],
  reporters: [ BasicReporter {} ] }
[log] <Bold><Cyan>üõ†Ô∏è Welcome to Nadle {version}!</Cyan></BoldDim>
[log] Using Nadle from /REPO_ROOT/node_modules/.pnpm/nadle{version}/node_modules/nadle/lib/index.js
[log] Loaded configuration from /ROOT/test/__fixtures__/main/nadle.configure.ts

[info] Using 1 worker for task execution
[info] Project directory: /ROOT/test/__fixtures__/main
[info] Resolved options: { cache: true,
  parallel: false,
  logLevel: 'info',
  showConfig: true,
  showSummary: false,
  excludedTasks: [],
  isWorkerThread: false,
  maxWorkers: 1,
  configPath:
   '/ROOT/test/__fixtures__/main/nadle.configure.ts',
  list: false,
  dryRun: false,
  stacktrace: false,
  minWorkers: 1,
  projectDir: '/ROOT/test/__fixtures__/main' }
[info] Detected environments: { CI: false, TEST: true }
[info] Execution started
[info] Resolved tasks: [  ]
[log] {
  "cache": true,
  "parallel": false,
  "logLevel": "info",
  "showConfig": true,
  "showSummary": false,
  "excludedTasks": [],
  "isWorkerThread": false,
  "maxWorkers": 1,
  "configPath": "/ROOT/test/__fixtures__/main/nadle.configure.ts",
  "list": false,
  "dryRun": false,
  "stacktrace": false,
  "minWorkers": 1,
  "projectDir": "/ROOT/test/__fixtures__/main"
}
[info] Execution finished
[log] 
<Bold><Green>RUN SUCCESSFUL</Green></BoldDim> in <Bold>{duration}</BoldDim>
[log] <Bold>0</BoldDim> task executed
`;

exports[`configure > can use configured options from config file > stdout 1`] = ``;
